linera-core/src/worker.rs

```diff
@@ -638,33 +642,82 @@ where
             }
         );
         certificate.check(committee)?;
+        let mut actions = NetworkActions::default();
+        let next_round = chain.manager.get().next_round();
         if chain
             .tip_state
             .get()
             .already_validated_block(block.height)?
-            || chain
-                .manager
-                .get_mut()
-                .check_validated_block(&certificate)?
-                == ChainManagerOutcome::Skip
+            || chain.manager.get().check_validated_block(&certificate)? == ChainManagerOutcome::Skip
         {
             // If we just processed the same pending block, return the chain info unchanged.
-            return Ok(ChainInfoResponse::new(&chain, self.key_pair()));
+            return Ok((ChainInfoResponse::new(&chain, self.key_pair()), actions));
         }
         if self
             .cache_recent_value(Cow::Borrowed(&certificate.value))
             .await
         {
-            let value = certificate.value.clone().into_confirmed();
-            self.cache_recent_value(Cow::Owned(value)).await;
+            if let Some(value) = certificate.value.clone().into_confirmed() {
+                self.cache_recent_value(Cow::Owned(value)).await;
+            }
```

Comment:
Seems like we could have a small helper function taking `&mut self` and a `&Value`.